src/main/java/net/fabricmc/fabric/mixin/client/rendercache/MixinBlockModelRenderer.java

```diff
@@ -0,0 +1,113 @@
+import net.minecraft.client.render.model.BakedModel;
+import net.minecraft.client.render.model.BakedQuad;
+import net.minecraft.util.crash.CrashException;
+import net.minecraft.util.crash.CrashReport;
+import net.minecraft.util.crash.CrashReportElement;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.world.ExtendedBlockView;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Random;
+
+@Mixin(BlockModelRenderer.class)
+public class MixinBlockModelRenderer {
+	private static final Direction[] fabric_directionValues = Direction.values();
+
+	// TODO NORELEASE: should this be split into two hooks to arrow overriding AO check?
+	@Inject(at = @At("HEAD"), method = "tesselate", cancellable = true)
+	public void tesselateSmoothRedir(ExtendedBlockView view, BakedModel model, BlockState state, BlockPos pos, BufferBuilder builder, boolean bool, Random random, long seed, CallbackInfoReturnable<Boolean> info) {
+		if (model instanceof DynamicBakedModel && view instanceof RenderCacheView) {
+			boolean useAO = MinecraftClient.isAmbientOcclusionEnabled() && state.getLuminance() == 0 && model.useAmbientOcclusion();
+
+			try {
+				Object renderData = ((DynamicBakedModel) model).getRenderData(state, (RenderCacheView) view, pos);
+
+				boolean result;
+				if (useAO) {
+					result = fabric_tesselateSmoothDynamic(renderData, view, (DynamicBakedModel) model, state, pos, builder, bool, random, seed);
+				} else {
+					result = fabric_tesselateFlatDynamic(renderData, view, (DynamicBakedModel) model, state, pos, builder, bool, random, seed);
+				}
+				info.setReturnValue(result);
+				info.cancel();
+			} catch (Throwable t) {
+				CrashReport report = CrashReport.create(t, "Tesselating dynamic block model");
+				CrashReportElement reportElementModel = report.addElement("Dynamic block being tesselated");
+				CrashReportElement.addBlockInfo(reportElementModel, pos, state);
+				reportElementModel.add("Using AO", useAO);
+				throw new CrashException(report);
+			}
+		}
+	}
+
+	public boolean fabric_tesselateSmoothDynamic(Object renderData, ExtendedBlockView view, DynamicBakedModel model, BlockState state, BlockPos pos, BufferBuilder builder, boolean allSides, Random rand, long seed) {
+		// TODO NORELEASE: AmbientOcclusionCalculator is non-public... ugh...
+		return fabric_tesselateFlatDynamic(renderData, view, model, state, pos, builder, allSides, rand, seed);
+	}
+
+	public boolean fabric_tesselateFlatDynamic(Object renderData, ExtendedBlockView view, DynamicBakedModel model, BlockState state, BlockPos pos, BufferBuilder builder, boolean allSides, Random rand, long seed) {
```

Comment:
You could simplify these patches by making renderData an attribute of RenderCacheView.  Its a bit of a hack, but RenderCacheView is always per-thread, and the only consumer of the render data will be the model we are rendering.

This in turn means you don't need to pass an additional parameter to tesselateSmooth or tesselateFlat - those calls can remain unchanged.  The only change in tesselate() would be check for a dynamic model, and populating the renderData attribute of the view.

In tesselateSmooth and tesselateFlat you'd only need a patch something like this:

```
@Redirect(method = "tesselateSmooth", at = @At(value = "INVOKE", 
            target = "Lnet/minecraft/client/render/model/BakedModel;getQuads(Lnet/minecraft/block/BlockState;Lnet/minecraft/util/math/Direction;Ljava/util/Random;)Ljava/util/List;"))
    private BlockPos onGetQuads(BakedModel model, @Nullable BlockState state, @Nullable Direction direction, Random rand)
    {
        return model instanceof DynamicBakedModel && view instanceof RenderCacheView 
            ? ((DynamicBakedModel)model).getQuads(((RenderCacheView)view).renderData(), state, direction, rand)
            : model.getQuads(state, direction, rand);
    }

```
Perhaps it should be an Inject instead of Redirect, but hopefully this conveys the idea.

Reply:
We can't do an Inject instead of a Redirect there, and I definitely don't want to stop other mods from also redirecting that particular call.

Reply:
I understand the hesitation to use a redirect, philosophically, but would invite you to reconsider in this specific instance.  The inject here is functionally a redirect of the entire method with a new implementation. It's still very invasive.

If mod authors want to implement finer-grained hooks that co-exist with the existing Fabric changes, they will need to reimplement the Fabric changes in their own method and inject with a higher priority.  Worse, that becomes the "normal" way to do anything with this method.

If you instead do a couple very narrow redirects on getQuads, authors can still make changes around those and the Fabric changes overall are less invasive.  You could also pass the BlockRenderLayer without a static ThreadLocal hack by attaching it to the RenderCacheView along with the RenderData.

Other use cases for redirecting that particular method seem unlikely.  An alternate render pipeline implementation will likely bypass tesselate() entirely, and I think that's fine.  Model interfaces can be extended for alternate render pipelines, but all models should still render with the vanilla pipeline. To this end, a model shouldn't know how it is being rendered to the screen.  It should simply produce requested vertex data per defined interfaces.

Reply:
>You could also pass the BlockRenderLayer without a static ThreadLocal hack by attaching it to the RenderCacheView along with the RenderData.

I can do that regardless. 

>If you instead do a couple very narrow redirects on getQuads, authors can still make changes around those and the Fabric changes overall are less invasive. You could also pass the BlockRenderLayer without a static ThreadLocal hack by attaching it to the RenderCacheView along with the RenderData.

Yes, but then we become the *only* party which can perform a narrow redirect on getQuads. The idea here is that if someone throws out, say, the vanilla processing for static block models - dynamic block models which require different provisions are unaffected and still use familiar logic. Besides, you *can* inject into methods *added* by a mixin, IIRC?